forecast/model.py

```diff
@@ -0,0 +1,304 @@
+    sequence_length : number of data points to use for prediction
+    """
+    def __init__(self, df, horizon=pd.Timedelta('1S'), interval=pd.Timedelta('1S'), sequence_length=5):
+        self.horizon = horizon
+        self.interval = interval
+        self.sequence_length = sequence_length
+        self.raw_df = df.resample(interval).sum()
+
+        self.set_transformers((None, None))
+
+    def __len__(self):
+        return self.X.shape[0]
+
+    def __getitem__(self, i):
+        if i >= self.sequence_length - 1:
+            i_start = i - self.sequence_length + 1
+            x = self.X[i_start:(i + 1)]
+        else:
+            padding = self.X[0].repeat(self.sequence_length - i - 1, 1)
+            x = self.X[0:(i + 1), :]
+            x = torch.cat((padding, x), 0)
+
+        return x, self.y[i].reshape((1, -1))
+    
+    def get_y_timestamp(self, ind):
+        return self.raw_df.index[ind] + self.horizon
+
+    def set_transformers(self, transformers):
+        x_transformer, y_transformer = transformers
+        shifted = self.raw_df.shift(
+            freq=-self.horizon).reindex_like(self.raw_df).ffill()
+
+        if (x_transformer == None or y_transformer == None):
+            self.X = torch.FloatTensor(self.raw_df.values)
+            self.y = torch.FloatTensor(shifted.values)
+            return
+
+        self.X = torch.FloatTensor(x_transformer.transform(self.raw_df.values))
+        self.y = torch.FloatTensor(y_transformer.transform(shifted.values))
+
+
+class ForecastModel(ABC):
+    """
+    Interface for all the forecasting models
+    """
+
+    def __init__(self, horizon, interval, sequence_length):
+        self._x_transformer = None
+        self._y_transformer = None
+
+        self.horizon = horizon
+        self.interval = interval
+        self.sequence_length = sequence_length
+
+    @property
+    def name(self):
+        return self.__class__.__name__
+
+    def fit(self, train_seqs: ForecastDataset) -> None:
+        """
+        Fit the model with sequences
+        :param train_seqs: List of training sequences and the expected output label in a certain horizon
+        :return:
+        """
+
+        assert self.horizon == train_seqs.horizon
+        assert self.interval == train_seqs.interval
+        assert self.sequence_length == train_seqs.sequence_length
+
+        self._x_transformer, self._y_transformer = self._get_transformers(
+            train_seqs.raw_df.values)
+        
+        transformed = copy.deepcopy(train_seqs)
+        transformed.set_transformers((self._x_transformer, self._y_transformer))
+
+        self._do_fit(transformed)
+
+    @abstractmethod
+    def _do_fit(
+            self, trains_seqs: ForecastDataset) -> None:
+        """
+        Perform fitting.
+        Should be overloaded by a specific model implementation.
+        :param train_seqs: List of training sequences and the expected output label in a certain horizon. Normalization
```

Comment:
I think this documentation came directly from the old NoisePage project -- might be a good time to take a fresh look at it.

1. Are we doing any normalization now?
2. "would have been done" -- by who, caller, callee?

Reply:
Will add this to the comments in the code. The data is normalized such that the training data is scaled to [-1,1]

Reply:
Cool, please mark resolved when fixed.